src/overlay121/DWWarp.c

```diff
@@ -0,0 +1,385 @@
+#include <nitro.h>
+#include <string.h>
+
+#include "core_sys.h"
+
+#include "struct_decls/struct_02006C24_decl.h"
+#include "struct_decls/sys_task.h"
+#include "struct_decls/struct_020203AC_decl.h"
+
+#include "struct_defs/struct_020170F4.h"
+#include "struct_defs/struct_02017248.h"
+#include "struct_defs/struct_02017294.h"
+#include "struct_defs/struct_0207C690.h"
+#include "struct_defs/struct_02099F80.h"
+#include "overlay115/camera_angle.h"
+
+#include "unk_02002328.h"
+#include "unk_02005474.h"
+#include "overlay_manager.h"
+#include "narc.h"
+#include "unk_0200D9E8.h"
+#include "unk_0200F174.h"
+#include "unk_020170BC.h"
+#include "unk_02017728.h"
+#include "heap.h"
+#include "unk_0201E3D8.h"
+#include "gx_layers.h"
+#include "unk_02020020.h"
+#include "unk_0202419C.h"
+#include "unk_02024220.h"
+#include "overlay121/DWWarp.h"
+
+typedef struct {
+    GenericPointerData * p3DCallback;
+    UnkStruct_020203AC * camera;
+    SysTask * task;
+    int frame_cnt;
+    int soundEffect_cnt;
+    UnkStruct_02017294 animation_obj;
+    UnkStruct_02017248 animation_animation;
+    UnkStruct_02017248 animation_animation2;
+    UnkStruct_020170F4 animation_model;
+    u32 move_frame;
+    int camera_angle;
+    int camera_angle_calc;
+    fx32 camera_distance;
+    int camera_perspective;
+    NNSFndAllocator allocator;
+} DistortionWorld_Warp;
+
+static void DW_Update(SysTask * task, void * data);
+static void DW_VBlankIntr(void * data);
+static void DW_VramSetBank(void);
+static void DW_InitCamera(DistortionWorld_Warp * warp);
+static void DWWarp_CameraEnd(DistortionWorld_Warp * warp);
+static void DWWarp_InitModel(DistortionWorld_Warp * warp);
+static void DWWarp_DeleteModel(DistortionWorld_Warp * warp);
+static void Model3D_Update(DistortionWorld_Warp * warp);
+static GenericPointerData * DWWarp_Init3D(int heapId);
+static void DWWarp_Setup3D(void);
+static void DWWarp_Exit3D(GenericPointerData * param0);
+static void DWWarp_CameraMove(DistortionWorld_Warp * warp);
+
+// Initializes the Distortion World warp effect overlay.
+int DWWarp_Init (OverlayManager * ovy, int * seq)
+{
+    DistortionWorld_Warp * dww;
+
+    SetMainCallback(NULL, NULL);
+    DisableHBlank();
+    GXLayers_DisableEngineALayers();
+    GXLayers_DisableEngineBLayers();
+
+    GX_SetVisiblePlane(0);
+    GXS_SetVisiblePlane(0);
+    GX_SetVisibleWnd(GX_WNDMASK_NONE);
+    GXS_SetVisibleWnd(GX_WNDMASK_NONE);
+    G2_BlendNone();
+    G2S_BlendNone();
+
+    Heap_Create(3, 30, 0x50000);
```

Comment:
**polish:** There are constants for these values in `constants/heap.h`; use ones that exist and make new ones for those that don't.